web/api/v1/api.go

```diff
@@ -829,12 +833,15 @@ func respond(w http.ResponseWriter, data interface{}) {
 		Data:   data,
 	})
 	if err != nil {
+		level.Error(api.logger).Log("msg", "error marshalling json response", "err", err)
 		return
 	}
-	w.Write(b)
+	if n, err := w.Write(b); err != nil {
+		level.Error(api.logger).Log("msg", "error writing response", "n", n, "err", err)
+	}
```

Comment:
Is there a reason for sending the header before the body has been successfully encoded?  It seems like it would make more sense to encode the data first, then send the Content-Type (and potentially Content-Length as it will be known) just before calling Write.  The explicit `w.WriteHeader(http.StatusOK)` is technically not needed as `w.Write` will do that implicitly.

If this is done, then an encoding error could be handled by returning an error to the client.